File: course-schedule/shaun_bfs.cpp
/*
Runtime: 20 ms, faster than 100.00% of C++ online submissions for Course Schedule.
Memory Usage: 8.9 MB, less than 4.63% of C++ online submissions for Course Schedule.
shaun c++ from jz java
*/

class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<vector<int>> edges(numCourses);  //vector of next for each node
        vector<int> degree(numCourses,0);
        //int degree[numCourses] = {0};
        for (auto p: prerequisites) {
            //p0 <-- p1
            int p0 = p.first, p1 = p.second;
            ++degree[p0];
            edges[p1].push_back(p0);
        }
        queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (degree[i] == 0) q.push(i);
        }
        int count = 0;
        while (!q.empty()){
            int node = q.front(); q.pop();
            count++;
            for (int next: edges[node]) {
                --degree[next];
                if (degree[next] == 0) {
                    q.push(next);
                }
            }
        }
        return count == numCourses;
        
    }
};

File: clone-graph/shaun-bfs.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}

Runtime: 27 ms, faster than 72.85% of Java online submissions for Clone Graph.
Memory Usage: 40 MB, less than 5.88% of Java online submissions for Clone Graph.
*/


//shaun adapted from python bfs, 40% tests passed, wrong answer
//fix a typo, then accepted, beats 92.00%, same idea as jz
public class Solution {
    public Node cloneGraph(Node node){
        
        if (node == null) 
            return null; 
            
        //use bfs algorithm to traverse the graph and get all nodes.
        Set<Node> nodes = getNodes(node);
        
        //copy nodes, store the old->new mapping information in a hashmap
        Map<Node, Node> mapping = new HashMap<>(); 
        for (Node n : nodes) {
            mapping.put(n, new Node(n.val));
        }
        
        //copy neighbors(edges)
        for (Node n : nodes) {
            Node new_node = mapping.get(n);
            for (Node neighbor : n.neighbors) {
                Node new_neighbor = mapping.get(neighbor);
                new_node.neighbors.add(new_neighbor);
            }
        }
        
        return mapping.get(node);
    }

    //bfs traversal of graph return set of nodes    
    //return a set or list? set does not order, we want bfs order
    //Set<UndirectedGraphNode> getNodes(UndirectedGraphNode node) {  --> wrong answer
    Set<Node> getNodes(Node node) {
        Queue<Node> q = new ArrayDeque<>();
        q.offer(node);
        
        
        Set<Node> res = new HashSet<>();
        res.add(node);
        while (!q.isEmpty()) {
            Node head = q.poll();
            for (Node neighbor : head.neighbors) {
                if (!res.contains(neighbor)) {
                    res.add(neighbor);
                    q.offer(neighbor);
                }
            }
        }
        return res;
    }
}

